core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1864,6 +1877,22 @@ abstract class RDD[T: ClassTag](
   // From performance concern, cache the value to avoid repeatedly compute `isBarrier()` on a long
   // RDD chain.
   @transient protected lazy val isBarrier_ : Boolean = dependencies.exists(_.rdd.isBarrier())
+
+  /**
+   * Whether the RDD's computing function is idempotent. Idempotent means the computing function
+   * not only satisfies the requirement, but also produce the same output sequence(the output order
+   * can't vary) given the same input sequence. Spark assumes all the RDDs are idempotent, except
+   * for the shuffle RDD and RDDs derived from non-idempotent RDD.
+   */
```

Comment:
This will mean all rdd's which are directly or indirectly reading from an unsorted shuffle output are not 'idempotent'.

Reply:
yes, that is expected, unless the computing function sorts the input data. For this case, we can override the `isIdempotent`.